packages/contracts/src/contracts/current/forwarder/Forwarder.sol

```diff
@@ -0,0 +1,315 @@
+    ///      The caller is sent all NFT's from the fill of orders
+    ///      This function will revert unless all NFT's are purchased.
+    /// @param orders An array of Order struct containing order specifications.
+    /// @param signatures An array of Proof that order has been created by maker.
+    /// @param feeOrders An array of Order struct containing order specifications for fees.
+    /// @param feeSignatures An array of Proof that order has been created by maker for the fee orders.
+    /// @return Amounts filled and fees paid by maker and taker.
+    function buyNFTTokens(
+        Order[] memory orders,
+        bytes[] memory signatures,
+        Order[] memory feeOrders,
+        bytes[] memory feeSignatures)
+        payable
+        public
+    {
+        require(msg.value > 0, ERROR_INVALID_INPUT);
+        address token = readAddress(orders[0].takerAssetData, 1);
+        require(token == address(ETHER_TOKEN), ERROR_INVALID_INPUT);
+
+        ETHER_TOKEN.deposit.value(msg.value)();
+        marketBuyNFTTokens(orders, signatures, feeOrders, feeSignatures, msg.value);
+    }
+
+    /// @dev Buys the NFT tokens, performing fee abstraction if required and deducts a proportional fee to fee recipient.
+    ///      This function is payable and will convert all incoming ETH into WETH and perform the trade on behalf of the caller.
+    ///      The caller is sent all NFT's from the fill of orders. This function will revert unless all NFT's are purchased.
+    /// @param orders An array of Order struct containing order specifications.
+    /// @param signatures An array of Proof that order has been created by maker.
+    /// @param feeOrders An array of Order struct containing order specifications for fees.
+    /// @param feeSignatures An array of Proof that order has been created by maker for the fee orders.
+    /// @param feeProportion A proportion deducted off the incoming ETH and sent to feeRecipient. The maximum value for this
+    ///        is 1000, aka 10%. Supports up to 2 decimal places. I.e 0.59% is 59.
+    /// @param feeRecipient An address of the fee recipient whom receives feeProportion of ETH.
+    /// @return Amounts filled and fees paid by maker and taker.
+    function buyNFTTokensFee(
+        Order[] memory orders,
+        bytes[] memory signatures,
+        Order[] memory feeOrders,
+        bytes[] memory feeSignatures,
+        uint16  feeProportion,
+        address feeRecipient)
+        payable
+        public
+    {
+        require(msg.value > 0, ERROR_INVALID_INPUT);
+        address token = readAddress(orders[0].takerAssetData, 1);
+        require(token == address(ETHER_TOKEN), ERROR_INVALID_INPUT);
+        uint256 remainingTakerTokenAmount = deductFees(msg.value, feeProportion, feeRecipient);
+        ETHER_TOKEN.deposit.value(remainingTakerTokenAmount)();
+        marketBuyNFTTokens(orders, signatures, feeOrders, feeSignatures, remainingTakerTokenAmount);
+    }
+
+    function deductFees(
```

Comment:
Rename to `deductFee`, singular since there is only a single fee being paid.

Reply:
Let's rename this to `payAndDeductFee`, it wasn't clear to me that this method sent the fees. `deductFee` could also simply be the arithmetic deduction of the fee amount.